File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());
        ArrayList<Integer>[] arr = new ArrayList[N + 1];
        boolean[] visited = new boolean[N + 1];
        for(int i = 1; i <= N; i++) {
            arr[i] = new ArrayList<Integer>();
        }
        for(int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());
            int u = Integer.parseInt(st.nextToken());
            int v = Integer.parseInt(st.nextToken());
            arr[u].add(v);
            arr[v].add(u);
        }
        Queue<Integer> que = new LinkedList<>();
        int cnt = 0;
        for (int i = 1; i <= N; i++) {
            if (!visited[i]) {
                que.add(i);
                visited[i] = true;
                cnt++;
                while (!que.isEmpty()) {
                    int node = que.poll();
                    for (int num : arr[node]) {
                        if (!visited[num]) {
                            visited[num] = true;
                            que.add(num);
                        }
                    }
                }
            }
        }
        System.out.println(cnt);
    }
}

File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.js
class Queue {
  constructor() {
    this.arr = [];
    this.head = 0;
  }

  add(num) {
    this.arr.push(num);
  }

  poll() {
    return this.arr[this.head++];
  }

  isEmpty() {
    return this.head === this.arr.length;
  }
}

const [[N, M], ...input] = require('fs').readFileSync('/dev/stdin').toString().trim().split('\n').map(line => line.trim().split(' ').map(Number));
const arr = Array.from({length : N + 1}, () => []);
for(let i = 0; i < M; i++) {
  arr[input[i][0]].push(input[i][1]);
  arr[input[i][1]].push(input[i][0]);
}
const que = new Queue();
const visited = new Array(N + 1).fill(false);
let cnt = 0;
for(let i = 1; i <= N; i++) {
  if(!visited[i]) {
    que.add(i);
    visited[i] = true;
    cnt++;
    while(!que.isEmpty()) {
      const node = que.poll();
      for(let num of arr[node]) {
        if(!visited[num]) {
          visited[num] = true;
          que.add(num);
        }
      }
    }
  }
}
console.log(cnt);